samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -59,6 +61,7 @@
 @InterfaceStability.Evolving
 public class StreamProcessor {
   private final JobCoordinator jobCoordinator;
+  private final List<StreamProcessorLifeCycleAware> lifeCycleAwares = new ArrayList<>();
```

Comment:
Makes sense. I added the life cycle listeners as a collection to make it more general, in case some other class other than ApplicationRunner wants to listen to it. Since it's causing confusion, I changed it to be a single aware object.